model-sync-lib/src/main/kotlin/org/modelix/model/sync/ModelImporter.kt

```diff
@@ -0,0 +1,332 @@
+package org.modelix.model.sync
+
+import org.modelix.model.api.*
+import org.modelix.model.data.ModelData
+import org.modelix.model.data.NodeData
+import java.io.File
+
+class ModelImporter(private val root: INode, val stats: ImportStats? = null) {
+
+    fun import(jsonFile: File) {
+        require(jsonFile.exists())
+        require(jsonFile.extension == "json")
+
+        val data = ModelData.fromJson(jsonFile.readText())
+        import(data)
+    }
+    
+    fun import(data: ModelData) {
+        stats?.reset()
+
+        syncProperties(root, data.root) // root original id is required for following operations
+        val allExistingNodes = root.getDescendants(true).toList()
```

Comment:
The algorithm creates too many collections containing all nodes. For large models one collection is already bad for the scalability, but here we have five collections (`allExistingNodes`, `originalIdToSpec`, `originalIdToExisting`, `allNodes` and `originalIdToRef`) which all are about as big as the model.

I think you only need the `originalIdToExistingMap`, which gets updated while new nodes are added.